types/src/transaction/webauthn.rs

```diff
@@ -0,0 +1,693 @@
+///
+/// See <https://www.w3.org/TR/webauthn-3/#authenticatorassertionresponse>
+#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+pub struct PartialAuthenticatorAssertionResponse {
+    /// This attribute contains the raw signature returned from the authenticator.
+    /// NOTE: Many signatures returned from WebAuthn assertions are not raw signatures.
+    /// As an example, secp256r1 signatures are encoded as an [ASN.1 DER Ecdsa-Sig_value](https://www.w3.org/TR/webauthn-3/#sctn-signature-attestation-types)
+    /// If the signature is encoded, the client is expected to convert the encoded signature
+    /// into a raw signature before including it in the transaction
+    signature: WebAuthnSignature,
+    /// This attribute contains the authenticator data returned by the authenticator.
+    /// See [`AuthenticatorData`](passkey_types::ctap2::AuthenticatorData).
+    #[serde(with = "serde_bytes")]
+    authenticator_data: Vec<u8>,
+    /// This attribute contains the JSON byte serialization of [`CollectedClientData`](CollectedClientData) passed to the
+    /// authenticator by the client in order to generate this credential. The exact JSON serialization
+    /// MUST be preserved, as the hash of the serialized client data has been computed over it.
+    #[serde(with = "serde_bytes")]
+    client_data_json: Vec<u8>,
+}
+
+impl PartialAuthenticatorAssertionResponse {
+    pub fn new(
+        webauthn_signature: WebAuthnSignature,
+        authenticator_data: Vec<u8>,
+        client_data_json: Vec<u8>,
+    ) -> Self {
+        PartialAuthenticatorAssertionResponse {
+            signature: webauthn_signature,
+            authenticator_data,
+            client_data_json,
+        }
+    }
+
+    /// Uses BCS serialization
+    pub fn to_bytes(&self) -> Vec<u8> {
+        bcs::to_bytes(&self).expect("Only unhandleable errors happen here.")
+    }
+
+    pub fn signature(&self) -> &WebAuthnSignature {
+        &self.signature
+    }
+
+    pub fn signature_bytes(&self) -> Vec<u8> {
+        bcs::to_bytes(&self.signature).expect("Only unhandleable errors happen here.")
+    }
+
+    /// In our adaptation of WebAuthn, the `challenge` provided to `authenticatorGetAssertion`
+    /// is the SHA3-256 digest of the `RawTransaction`.
+    ///
+    /// This function should do the following:
+    /// 1. Verify `actual_challenge` and expected challenge from message are equal
+    /// 2. Construct `verification_data` as the binary concatenation of
+    ///    authenticator_data and SHA-256(client_data_json)
+    /// 3. Signature verification
+    ///
+    /// See WebAuthn §6.3.3 `authenticatorGetAssertion` for more info
+    pub fn verify<T: Serialize + CryptoHash>(
+        &self,
+        message: &T,
+        public_key: &AnyPublicKey,
+    ) -> Result<()> {
+        let collected_client_data: CollectedClientData =
+            serde_json::from_slice(self.client_data_json.as_slice())?;
```

Comment:
I am not a big fan of deserializing the signed message and then matching the TXN hash in one of the the deserialized struct's fields.

My main hesitation is that JSON deserialization bugs will lead to signature forgery attacks.

This is another reason, beyond smaller signature size, why I think the verifier should reassemble the signed message from its components (assuming canonical encodings). This should involve very simple string concatenation code & be way less bug prone.

cc @gedigi